core/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java

```diff
@@ -407,11 +407,10 @@ private ThreadContextStruct putHeaders(Map<String, String> headers) {
             if (headers.isEmpty()) {
                 return this;
             } else {
-                final Map<String, String> newHeaders = new HashMap<>();
+                final Map<String, String> newHeaders = new HashMap<>(this.requestHeaders);
```

Comment:
Alternatively, you can still initialize the `newHeaders` to `requestHeaders`, but change the call to `putSingleHeader` to `newHeaders.put`. Since `headers` and `requestHeaders` are already maps, there is no possibility for duplicate keys within themselves.